inspirehep/modules/multieditor/actions.py

```diff
@@ -0,0 +1,376 @@
+
+
+class Action(object):
+    def __init__(self, keypath, value=None, match_type=None, update_value=None,
+                 conditions=None):
+        """
+        :param keypath: path to value
+        :type keypath: array(string)
+
+       :param value: value to add or update to
+       :type value: string
+
+       :param match_type: type of matching
+       :type match_type: string
+
+       :param update_value: value to replace or delete
+       :type update_value: string
+
+       :param conditions: conditions of the action
+       :type conditions: array(object)
+        """
+        self.keypath = keypath
+        self.value = value
+        self.match_type = match_type
+        self.update_value = update_value
+        self.conditions = conditions
+        self.changed = False
+
+    def get_next_key_and_condition_check(self, record, schema, position, conditions_passed):
+        """
+
+        :param record: record to be processed
+        :param position: recursion level
+        :param conditions_passed: number of conditions that got checked and passed
+        :return: returns the new number of conditions passed,
+         the new key for this recursion and if a condition got checked and the failed the check
+        """
+        key = self.keypath[position]
+        condition_failed = False
+        if self.conditions:
+            for condition in self.conditions:
+                should_check = False
+                if position < len(condition.get('keypath')):
+                    if key != condition.get('keypath')[position] and\
+                     (position == 0 or self.keypath[position - 1] == condition.get('keypath')[position - 1]):
+                        should_check = True
+                    elif key == condition.get('keypath')[position] and position == len(self.keypath) - 1:
+                        should_check = True
+                    if should_check:
+                        if not check_value(record=record, schema=schema, keypath=condition.get('keypath'),
+                                           update_value=condition.get('value', ''),
+                                           match_type=condition.get('match_type', ''), position=position):
+                            condition_failed = True
+                        else:
+                            conditions_passed = conditions_passed + 1  # number of conditions that passed successfully
+        return conditions_passed, key, condition_failed
+
+
+class Addition(Action):
+    """Class that when applied adds a new primitive field or object to the selected path
+    if the conditions are satisfied"""
+    def apply(self, record, schema, position=0, conditions_passed=0):
+        """
+        :param record: record in which the field to be updated resides
+        :param schema: schema on which our record is based
+        :param position: level of recursion
+        :param conditions_passed: number of conditions that run and passed
+        :return:
```

Comment:
The action can still be used since im not creating a new instance of the class. Im just replacing the record to be processed in the existing one.Also in most cases I have no use of the old record and its more efficient for me when processing the records to have in the memory just the ones that are being processed.

Reply:
Exactly because you are not creating new instances, is why you can't reuse them, as once the `self.changed` is set, it will always be true. It's also only set in the `apply` method, so you don't really need a state to be passed around to other methods, or future calls, that's why it does not really fit as instance property, but more as return value of the `apply` method.
If what you want is to keep a flag about if any of the records processed so far has changed, I'd suggest changing the name of the class and the method, something like `ActionProcessor` for the basic one, `DeleteProcessor` for the delete, and then instead of `apply` should be `process`. That way it's clear that it's not applying an action to a record, but processing a set of records.

> Im just replacing the record to be processed in the existing one.Also in most cases I have no use of the old record and its more efficient for me when processing the records to have in the memory just the ones that are being processed.

I'm not sure what you mean with this. I don't see how passing the changed as return value modifies the way you process records. For example, in the `process_records` task, instead of:
```python
    for record in records:
        for action in actions:
            action.apply(record=record, schema=schema)
            if action.changed:
                commit_record = True
                action.changed = False
```
you'll have:
```python
    for record in records:
        for action in actions:
            if action.apply(record=record, schema=schema):
                commit_record = True
```

Reply:
Ah okay got it! Thanks :)